src/main/groovy/wooga/gradle/unity/tasks/GenerateUpmPackage.groovy

```diff
@@ -120,11 +153,25 @@ class GenerateUpmPackage extends Tar implements BaseSpec {
             }
             slurper.parse(it.asFile)["name"].toString()
         })
+
+        Provider<String> packageVersionOnFile = packageManifestFile.map({
+            def slurper = new JsonSlurper()
+            if (!it.asFile.exists()) {
+                return null
+            }
+            slurper.parse(it.asFile)["version"].toString()
+        })
+
         packageName.convention(packageNameOnFile)
+        archiveVersion.set(packageVersionOnFile)
         archiveBaseName.set(packageName)
+        preserveFileTimestamps = false
+        reproducibleFileOrder = true
+
         filesMatching(packageManifestFileName) {
-            filter { it.replaceAll(/"name" : ".*?"/, "\"name\" : \"${packageName.get()}\"") }
-            filter { it.replaceAll(/"version" : ".*?"/, "\"version\" : \"${archiveVersion.get()}\"") }
+            if(it.getFile().absolutePath.startsWith(packageDirectory.get().asFile.absolutePath)) {
+                it.exclude()
```

Comment:
Just so I don't die stupid, now i want to understand this groovy crazyness. 
So basically you're calling exclude() on the iterator. Is the iterator a file? How does that end up telling the GenerateUpmPackage Tar Task to not include that file in the tar ?

Reply:
also, why does GenerateUpmPackage implement BaseSpec ? Isn't BaseSpec the base class for unit tests in groovy ?

Reply:
Ok some termonology here. A spec is short for specification. In test terms like in spock it means a specification how an object/class should work. It is behavior testing. A `copy spec` in gradle is a specification which describes how to copy a set of files. The tar task and any task that creates an archive are exposing a copy spec since they take a bunch of files and copy them into an archive.

What I do here is to define two from sources.
1. the package directory
2. a temp directory where the package.json got copied and modified

The line which confuses you works like this:
* the copy action will walk over the input files.
* I registed a closure to be called for each file info
* The input of the closure is an object that has pointers to the source file and the destination file
* it is up to the user if he wants to do some operations during the copy.
    * renaming
    * changeing content
    * or skipping that file for whatever reason
* I skip the package.json file when it originates from the package directory. I'm only interested in the *other* package.json from the temp directory.

`it` in this case is not the iterator. It is the current copy descripor represented by [FileCopyDetails](https://docs.gradle.org/current/javadoc/org/gradle/api/file/FileCopyDetails.html).
For more infos check [here](https://docs.gradle.org/current/javadoc/org/gradle/api/file/CopySpec.html#eachFile-groovy.lang.Closure-) and [here](https://docs.gradle.org/current/javadoc/org/gradle/api/file/FileCopyDetails.html)